### PR TITLE
Cleanup Any* wrappers related code

### PR DESCRIPTION
**Pull request by romac**
See: #218

## Description

This PR improves the ergonomics of the way we deal with multiple client types at runtime,
ie. the various `Any*` wrappers and the associated traits.

- Removes uses of `dyn ClientState` and `dyn ConsensusState`
- Move validation functions out of the `ClientState` trait and into the `ClientDef` trait
- Add method to get a `AnyClient` from a `ClientType`
- Remove type parameter on messages structs, enums, `process` and `keep` functions
______

For contributor use:

- [ ] Unit tests written
- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by romac** [comment]
@ancazamfir @adizere I wonder what you think of this item?

> - Move validation functions out of the ClientState trait and into the ClientDef trait

To me, it seemed more appropriate to have this live in the `ClientDef` implementation as this is client-specific (ie. the client state is mostly data, but it's the client definition that defines how to validate the client state).

I don't feel strongly either way though, thoughts welcome!

**Review by adizere — COMMENTED** [review]
Didn't finish the review. Just submitting very basic comments here until I manage to compile the branch..

**Comment by adizere** [comment]
> > * Move validation functions out of the ClientState trait and into the ClientDef trait
> 
> To me, it seemed more appropriate to have this live in the `ClientDef` implementation as this is client-specific (ie. the client state is mostly data, but it's the client definition that defines how to validate the client state).

Agree. Your suggestion makes sense and I prefer it to the older solution where ClientState was overloaded. I think, however, the older solution was that way because of the influence of [ICS02](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics#definitions), which states:

> ClientState is an opaque type representing the state of a client. A ClientState __must expose__ query functions to verify membership or non-membership of key/value pairs in state at particular heights and to retrieve the current ConsensusState.

That being said, if it affords us having this cleaner implementation, I would vote that it's ok to depart a bit from ICS02 and go forward with the suggestion in this branch.

File: modules/src/ics02_client/client_def.rs [hunk 1]
```diff
@@ -216,8 +167,152 @@ pub enum AnyClient {
...
+        match client_type {
+            ClientType::Tendermint => Self::Tendermint(TendermintClient),
+
+            #[cfg(test)]
+            ClientType::Mock => Self::Mock(MockClient),
+        }
+    }
+}
+
+// ⚠️  Beware of the awful boilerplate below ⚠️
```
Comment (brapse) [inline modules/src/ics02_client/client_def.rs:181]: cuuuuute

**Comment by romac** [comment]
@adizere 

> I'm not able to compile locally.

I suspect you need to update your version of Rust. You can do so by running:

```
$ rustup update
```

Indeed, `Option::zip` has only been [stabilized](https://doc.rust-lang.org/stable/src/core/option.rs.html#931) since Rust v1.46.0.

My bad, I should have mentioned this in the PR. For my defense, it's sometimes hard to keep track of new APIs and their stabilization version.

File: modules/src/lib.rs [hunk 2]
```diff
@@ -9,7 +9,7 @@
     unused_qualifications,
     rust_2018_idioms
 )]
-#![allow(dead_code)]
+#![allow(dead_code, unused_imports, unused_variables)]
```
[the remark this answers is not in the document]
Comment (romac) [reply modules/src/lib.rs:12]: Right, forgot to remove them after the refactoring.

**Comment by romac** [comment]
> Your suggestion makes sense and I prefer it to the older solution where ClientState was overloaded.

Yeah that was my feeling as well, hence the departure from ICS 02.

> That being said, if it affords us having this cleaner implementation, I would vote that it's ok to depart a bit from ICS02 and go forward with the suggestion in this branch.

It's actually orthogonal to the rest of the PR, just a matter of which trait these functions should live in. So if we feel we'd rather stick to the spec more closely, we can always move them back into the `ClientState` trait.

**Review by brapse — APPROVED** [review]
Great to have this moving forward. A bit weary of the roll your own inheritance system here but I can definitely appreciate the difficulties of supporting an unknown number of clients at runtime. Might be something we circle back on later but looks for as  next step 👍

File: modules/src/ics02_client/client_def.rs [hunk 3]
```diff
@@ -216,8 +167,152 @@ pub enum AnyClient {
...
+        proof: &CommitmentProof,
+        client_id: &ClientId,
+        consensus_height: Height,
+        expected_consensus_state: &AnyConsensusState,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        match self {
+            Self::Tendermint(client) => {
+                let (client_state, expected_consensus_state) = downcast!(
+                    client_state => AnyClientState::Tendermint,
+                    expected_consensus_state => AnyConsensusState::Tendermint,
```
Comment (ancazamfir) [inline modules/src/ics02_client/client_def.rs:242]: The `expected_consensus_state` should be of the same type as the local chain (`Tendermint` in production or `Mock` in testing). This is different than the `ClientDef` type. For example`self` could be a client for `Substrate` while the `expected_consensus_state` stored by "our" client on the substrate chain is `Tendermint`. Not entirely sure of the best way to get the local chain info. One possibility is to pass the `Reader` context as a parameter and get the local chain type from there. Didn't have a chance to review this before it was merged, only looked at it now while merging master to `anca/proto_update`. Here we have a similar problem for the new `verify_full_client_state()` method. Let's try to take care of this in https://github.com/informalsystems/ibc-rs/pull/227

**Pull request merged**